src/aws_encryption_sdk/materials_managers/caching.py

```diff
@@ -111,10 +100,21 @@ def __attrs_post_init__(self):
         if self.max_age <= 0.0:
             raise ValueError("max_age cannot be less than or equal to 0")
 
+        options_provided = [
+            option is not None for option in (self.backing_materials_manager, self.keyring, self.master_key_provider)
+        ]
+        provided_count = len([is_set for is_set in options_provided if is_set])
+
+        if provided_count != 1:
+            raise TypeError("Exactly one of 'backing_materials_manager', 'keyring', or 'key_provider' must be provided")
```

Comment:
Is TypeError the best exception we can throw here? Is there a better custom exception class, or should we create one?

Reply:
`TypeError` is consistent with what built-ins and `attrs` throw if you provide incorrect parameters or types.

```python
>>> def foo(b):
...     print("foo")
... 
>>> foo(3)
foo
>>> foo(b=3)
foo
>>> foo(f=3)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() got an unexpected keyword argument 'f'
>>> foo(3, 4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: foo() takes 1 positional argument but 2 were given

```